static/app/views/dashboardsV2/gridLayout/dashboard.tsx

```diff
@@ -280,8 +288,13 @@ const GridItem = styled('div')`
   }
 `;
 
-export function generateWidgetId(widget: Widget, index: number) {
-  return widget.id ? `${SAVED_WIDGET_PREFIX}-${widget.id}` : `index-${index}`;
+export function generateGridItemId(widget: Widget) {
+  const id = widget.id ?? widget.tempId;
+  if (id) {
+    return `${WIDGET_PREFIX}-${id}`;
+  }
+
+  return Date.now().toString();
```

Comment:
This function seems like it's serving 2 concerns right now:
- constructing keys for the grid items 
- generating a unique `tempId` for newly added widgets

I would recommend separating the logic for these two if possible. ie on line 150 of this file (`dashboard.tsx`), maybe update the line to something like `{...widget, tempId: Date.now().toString()},` since this is the only area where a new `tempId` is generated. Then you can refactor `generateGridItemId` to something like this since `generateGridItemId` would not longer need be responsible for generating `tempId`s. 
```
export function constructGridItemKey(widget: Widget) {
  return `${WIDGET_PREFIX}-${widget.id ?? widget.tempId}`;
}
```

Reply:
Also not an issue, but instead of using `Date.now().toString()`, you could try using a UUID instead since UUIDs seem to fit this use case pretty well. `Date.now().toString()` seems to work fine though so no need to switch to UUIDs if we don't have an easy way of generating one right now.

Reply:
So I kept it as `Date.now().toString()` because it's a very short lived thing and we don't have a specified UUID library.

Great observation for noticing the 2 concerns though. It made me feel uncomfortable too because it kept nagging at me that the code branches off depending on the context of its usage (i.e. called when assigning a tempId vs called when rendering widgets). Thank you for that.